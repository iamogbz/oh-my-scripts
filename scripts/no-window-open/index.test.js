const { toMatchDiffSnapshot } = require("snapshot-diff");
expect.extend({ toMatchDiffSnapshot });

const spyConsoleLog = jest
  .spyOn(console, "log")
  .mockImplementation(() => undefined);
const spyWindowOpen = jest.spyOn(window, "open");
require("./index.user");

describe("no-window-open", () => {
  beforeAll(jest.useFakeTimers);
  afterAll(jest.useRealTimers);

  it("blocks window open", () => {
    const url = "https://example.com";
    expect(window.open(url)).toMatchSnapshot();
    expect(spyWindowOpen).not.toHaveBeenCalled();
    expect(spyConsoleLog).toHaveBeenCalledWith(expect.stringContaining(url));
  });

  it("shows popup after time", () => {
    const url = "https://example.com";
    window.open(url);
    expect(document.body).toMatchSnapshot();
    const prevBody = document.body.cloneNode(true);
    jest.advanceTimersByTime(300);
    expect(prevBody).toMatchDiffSnapshot(document.body);
  });

  it("hides popup after timer run", () => {
    const url = "https://example.com";
    window.open(url);
    jest.advanceTimersByTime(300);
    const prevBody = document.body.cloneNode(true);
    jest.runOnlyPendingTimers();
    expect(prevBody).toMatchDiffSnapshot(document.body);
  });

  it("sets the url of the popup link via window location", () => {
    const dummyWindow = window.open("https://example.com");
    const getUrl = () =>
      document.body
        .querySelector("#iamogbz-no-window-open-popup-element-link")
        .getAttribute("href");
    expect(getUrl()).toEqual("https://example.com");

    dummyWindow.location = "https://example.com/2";
    expect(getUrl()).toEqual("https://example.com/2");

    dummyWindow.location.href = "https://example.com/3";
    expect(getUrl()).toEqual("https://example.com/3");

    dummyWindow.location.assign("https://example.com/4");
    expect(getUrl()).toEqual("https://example.com/4");

    dummyWindow.location.replace("https://example.com/5");
    expect(getUrl()).toEqual("https://example.com/5");
  });
});
