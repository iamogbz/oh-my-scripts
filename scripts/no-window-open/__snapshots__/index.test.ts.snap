// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no-window-open blocks window open 1`] = `
Object {
  "location": Object {
    "assign": [Function],
    "href": "https://example.com",
    "replace": [Function],
  },
}
`;

exports[`no-window-open delays hides popup after timer run on mouseover 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,8 +1,8 @@
  <body>
    <div
-     class=\\"iamogbz-no-window-open-popup-element-visible\\"
+     class=\\"\\"
      id=\\"iamogbz-no-window-open-popup-element\\"
    >
      <div>
        This page just attempted to open a url.
  Click on it below to proceed with navigation."
`;

exports[`no-window-open hides popup after timer run 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,8 +1,8 @@
  <body>
    <div
-     class=\\"iamogbz-no-window-open-popup-element-visible\\"
+     class=\\"\\"
      id=\\"iamogbz-no-window-open-popup-element\\"
    >
      <div>
        This page just attempted to open a url.
  Click on it below to proceed with navigation."
`;

exports[`no-window-open shows popup after time 1`] = `
<body>
  <div
    id="iamogbz-no-window-open-popup-element"
  >
    <div>
      This page just attempted to open a url.
Click on it below to proceed with navigation.
    </div>
    <a
      href="https://example.com"
      id="iamogbz-no-window-open-popup-element-link"
      referrer="noreferrer"
      target="_blank"
    >
      https://example.com
    </a>
    <button
      id="iamogbz-no-window-open-popup-element-close"
    >
      Dismiss
    </button>
  </div>
</body>
`;

exports[`no-window-open shows popup after time 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,7 +1,8 @@
  <body>
    <div
+     class=\\"iamogbz-no-window-open-popup-element-visible\\"
      id=\\"iamogbz-no-window-open-popup-element\\"
    >
      <div>
        This page just attempted to open a url.
  Click on it below to proceed with navigation."
`;
